### PR TITLE
feat: remove `Boxed` from `Channel.flix`

### PR DESCRIPTION
**Pull request by JonathanStarup**
Related to #2169.

Based on disucssion in #2696

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Overall looks good. I will propose some stylistic changes.

Question question: Why do you need eq and order? I think they are dangerous.

**Comment by JonathanStarup** [comment]
> Overall looks good. I will propose some stylistic changes.
> 
> Question question: Why do you need eq and order? I think they are dangerous.

locking order. the functions could also be inlined but then its not so easy to call `Array.sort`.

**Comment by magnus-madsen** [comment]
> > Overall looks good. I will propose some stylistic changes.
> > Question question: Why do you need eq and order? I think they are dangerous.
> 
> locking order. the functions could also be inlined but then its not so easy to call `Array.sort`. Its only on the private admin enum not on the mpmc

What about using Array.sortBy?

**Comment by JonathanStarup** [comment]
> What about using Array.sortBy?

Of yeah that works. Why are the instances dangerous?

**Comment by magnus-madsen** [comment]
Actually they are not because you only refer to immutable data 👍 

I had not though about that.

Even so, I would prefer a few well-chosen sortBy.

File: main/src/library/Concurrent/Channel.flix [hunk 1]
```diff
@@ -67,59 +72,50 @@ namespace Concurrent/Channel {
     /// conditions in `waitingPutters` comes from locks created by the get
     /// function or the select function.
     ///
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:74]: As always, I would ask for some documentation here to explain that Mpmc now consists of two parts.

File: main/src/library/Concurrent/Channel.flix [hunk 2]
```diff
@@ -67,59 +72,50 @@ namespace Concurrent/Channel {
...
-    /// Order is based on the channel `id` and is used by `select` to lock
-    /// in a global ordering.
-    ///
-    instance Order[Mpmc] {
-        pub def compare(x: Mpmc, y: Mpmc): Comparison =
-        let Mpmc(a_id, _, _, _, _, _, _) = x;
-        let Mpmc(b_id, _, _, _, _, _, _) = y;
-        a_id <=> b_id
-    }
+    enum MpmcAdmin(
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:86]: I wish we had a better name, but I don't have any suggestions.

File: main/src/library/Concurrent/Channel.flix [hunk 3]
```diff
@@ -171,30 +169,71 @@ namespace Concurrent/Channel {
         unlock(channelLock);
         element
 
+    ///
+    /// Receives an element from the channel `c`. This will crasah if the
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:173]: crash

File: main/src/library/Concurrent/Channel.flix [hunk 4]
```diff
@@ -67,59 +72,50 @@ namespace Concurrent/Channel {
...
-    /// Equality is based on the channel `id` and is needed for the `Order`
-    /// instance.
-    ///
-    instance Eq[Mpmc] {
-        pub def eq(x: Mpmc, y: Mpmc): Bool =
-        let Mpmc(a_id, _, _, _, _, _, _) = x;
-        let Mpmc(b_id, _, _, _, _, _, _) = y;
-        a_id == b_id
+    /// Returns the `MpmcAdmin` of `c`.
+    pub def mpmcAdmin(c: Mpmc[a]): MpmcAdmin = {
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:81]: You have this function, but don't use it consistently. Perhaps drop it?

File: main/src/library/Concurrent/Channel.flix [hunk 5]
```diff
@@ -171,30 +169,71 @@ namespace Concurrent/Channel {
...
+    /// Receives an element from the channel `c`. This will crasah if the
+    /// channel is empty.
+    ///
+    @Internal
+    pub def assertiveGet(c: Mpmc[a]): a \ IO = {
+        let Mpmc(admin, elementDeque) = c;
+        let MpmcAdmin(_, channelLock, _, _, size, _, _) = admin;
+        lock(channelLock);
+
+        // Try to get the element (which could already be taken by someone
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:182]: Lets phrase this as: Gets an element from the channel `c` which must be non-empty.

File: main/src/library/Concurrent/Channel.flix [hunk 6]
```diff
@@ -171,30 +169,71 @@ namespace Concurrent/Channel {
         unlock(channelLock);
         element
 
+    ///
+    /// Receives an element from the channel `c`. This will crasah if the
+    /// channel is empty.
+    ///
+    @Internal
+    pub def assertiveGet(c: Mpmc[a]): a \ IO = {
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:177]: How about `unsafeGet`?

File: main/src/library/Concurrent/Channel.flix (hunk 3 above)
Comment (magnus-madsen) [reply main/src/library/Concurrent/Channel.flix:173]: "Assumes the channel to be non-empty".

File: main/src/library/Concurrent/Channel.flix [hunk 7]
```diff
@@ -171,30 +169,71 @@ namespace Concurrent/Channel {
...
+        let MpmcAdmin(_, channelLock, _, _, size, _, _) = admin;
+        lock(channelLock);
+
+        // Try to get the element (which could already be taken by someone
+        // else)
+        let optionalElement = MutDeque.popFront(elementDeque);
+        size := deref size - 1;
+
+        match optionalElement {
+            case None => bug!("assertiveGet could not retrieve channel element")
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:188]: Explain this cannot happen.

File: main/src/library/Concurrent/Channel.flix [hunk 8]
```diff
@@ -240,44 +281,30 @@ namespace Concurrent/Channel {
...
-        let element = MutDeque.popFront(elementDeque);
-
-        // If there was an element, signal waiting setters
-        Option.foreach(_ -> signalPutters(c), element);
-
-        unlock(channelLock);
-        element
-
     ///
     /// Recursive helper function for select, returns the first channel that
```
Comment (magnus-madsen) [inline main/src/library/Concurrent/Channel.flix:289]: Docs.

**Review by magnus-madsen — COMMENTED** [review]
All my comments are nitpicks; I am just asking for better documentation and naming.

**Comment by magnus-madsen** [comment]
(This implementation could be an interesting target for verification.)

File: main/src/library/Concurrent/Channel.flix (hunk 4 above)
Comment (JonathanStarup) [reply main/src/library/Concurrent/Channel.flix:81]: It was meant for lowering (`pub def`), It seems annoying to unpack the ADT in the compiler

Comment (JonathanStarup) [reply main/src/library/Concurrent/Channel.flix:81]: I also though about making the admin a record type since it could be nice to access fields directly

Comment (magnus-madsen) [reply main/src/library/Concurrent/Channel.flix:71]: I think we can afford to use some records, but that could be in another iteration.

**Comment by magnus-madsen** [comment]
Other than the build, I think we could try to merge this.

**Pull request merged**